libkbfs/chat_local.go

```diff
@@ -6,74 +6,247 @@ package libkbfs
 
 import (
 	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
 
 	"github.com/keybase/client/go/logger"
 	"github.com/keybase/client/go/protocol/chat1"
+	"github.com/keybase/kbfs/kbfscrypto"
 	"github.com/keybase/kbfs/tlf"
+	"github.com/pkg/errors"
 )
 
-// ChatLocal is a local implementation for chat.
-//
-// TODO: fill in the logic to allow for testing.
-type ChatLocal struct {
+type convLocal struct {
+	convID   chat1.ConversationID
+	chanName string
+	messages []string
+	cbs      []ChatChannelNewMessageCB
+	mtime    time.Time
+}
+
+type convLocalByIDMap map[string]*convLocal
+
+type convLocalByTypeMap map[tlf.Type]convLocalByIDMap
+
+type convLocalByNameMap map[tlf.CanonicalName]convLocalByTypeMap
```

Comment:
Why did you choose to do name -> type -> ID instead of type -> name -> ID? Seems better for locality to go in order of increasing cardinality.

Reply:
It was arbitrary.  I'll change it.